src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/RazorFileChangeDetector.cs

```diff
@@ -130,11 +146,74 @@ protected virtual IReadOnlyList<string> GetExistingRazorFiles(string workspaceDi
+            lock (_pendingNotificationsLock)
+            {
+                if (!_pendingNotifications.TryGetValue(physicalFilePath, out var currentNotification))
+                {
+                    currentNotification = new DelayedFileChangeNotification();
+                    _pendingNotifications[physicalFilePath] = currentNotification;
+                }
+
+                if (currentNotification.ChangeKind != null)
+                {
+                    // We've already has a file change event for this file. Chances are we need to normalize the result.
+
+                    Debug.Assert(currentNotification.ChangeKind == RazorFileChangeKind.Added || currentNotification.ChangeKind == RazorFileChangeKind.Removed);
+
+                    if (currentNotification.ChangeKind != kind)
+                    {
+                        // Previous was added and current is removed OR previous was removed and current is added. Either way there's no
+                        // actual change to notify, null it out.
+                        currentNotification.ChangeKind = null;
+                    }
+                    else
+                    {
+                        Debug.Fail($"Unexpected {kind} event because our prior tracked state was the same.");
+                    }
+                }
+                else
+                {
+                    currentNotification.ChangeKind = kind;
+                }
+
+                if (currentNotification.NotifyTask == null)
+                {
+                    // The notify task is only ever null when it's the first time we're being notified about a change to the corresponding file.
+                    currentNotification.NotifyTask = NotifyAfterDelayAsync(physicalFilePath);
+                }
+            }
+        }
+
+        private async Task NotifyAfterDelayAsync(string physicalFilePath)
         {
-            Task.Factory.StartNew(
-                () => FileSystemWatcher_RazorFileEvent(physicalFilePath, kind),
-                CancellationToken.None, TaskCreationOptions.None, _foregroundDispatcher.ForegroundScheduler);
+            await Task.Delay(EnqueueDelay).ConfigureAwait(false);
+
+            OnStartingDelayedNotificationWork();
+
+            lock (_pendingNotificationsLock)
+            {
+                var result = _pendingNotifications.TryGetValue(physicalFilePath, out var notification);
+                Debug.Assert(result, "We should always have an associated notification after delaying an update.");
+
+                _pendingNotifications.Remove(physicalFilePath);
+
+                if (notification.ChangeKind == null)
+                {
+                    // The file to be notified has been brought back to its original state.
+                    // Aka Add -> Remove is equivalent to the file never having been added.
+
+                    OnNoopingNotificationWork();
+
+                    return;
+                }
+
+                _ = Task.Factory.StartNew(
```

Comment:
Why put the value on the throw-away variable instead of just not assigning it? Trying to communicate that we know this returns a value and don't care?

Reply:
Great question, fixed that in my latest PR: https://github.com/dotnet/aspnetcore-tooling/pull/1819